src/Microsoft.ML.Data/DataLoadSave/PartitionedFileLoader.cs

```diff
@@ -0,0 +1,726 @@
+                };
+
+                _columns = _columns.Concat(new[] { pathCol }).ToArray();
+            }
+
+            Schema = CreateSchema(_host, _columns, _subLoader);
+        }
+
+        private PartitionedFileLoader(IHost host, ModelLoadContext ctx, IMultiStreamSource files)
+        {
+            Contracts.AssertValue(host);
+            _host = host;
+            _host.AssertValue(ctx);
+            _host.AssertValue(files);
+
+            // ** Binary format **
+            // int: tailing directory count
+            // int: number of columns
+            // foreach column:
+            //   string: column representation
+            // string: subloader
+            // model: file path spec
+
+            _tailingDirCount = ctx.Reader.ReadInt32();
+
+            int numColumns = ctx.Reader.ReadInt32();
+            _host.CheckDecode(numColumns >= 0);
+
+            _columns = new Column[numColumns];
+            for (int i = 0; i < numColumns; i++)
+            {
+                var column = Column.Parse(ctx.LoadString());
+                _host.CheckDecode(column != null);
+                _columns[i] = column;
+            }
+
+            var loader = SubComponent.Parse(ctx.LoadString());
+            _subLoader = new SubComponent<IDataLoader, SignatureDataLoader>(loader.Kind, loader.Settings);
+
+            ctx.LoadModel<IPartitionedPathParser, SignatureLoadModel>(_host, out _pathParser, "FilePathSpec");
+
+            _files = files;
+            Schema = CreateSchema(_host, _columns, _subLoader);
+        }
+
+        public static PartitionedFileLoader Create(IHostEnvironment env, ModelLoadContext ctx, IMultiStreamSource files)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            IHost host = env.Register(RegistrationName);
+
+            env.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel(GetVersionInfo());
+            env.CheckValue(files, nameof(files));
+
+            return host.Apply("Loading Model",
+                ch => new PartitionedFileLoader(host, ctx, files));
+        }
+
+        public void Save(ModelSaveContext ctx)
+        {
+            Contracts.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            // ** Binary format **
+            // int: tailing directory count
+            // int: number of columns
+            // foreach column:
+            //   string: column representation
+            // string: subloader
+            // model: file path spec
+
+            ctx.Writer.Write(_tailingDirCount);
+
+            ctx.Writer.Write(_columns.Length);
+            StringBuilder sb = new StringBuilder();
+            foreach (var col in _columns)
+            {
+                sb.Clear();
+                _host.Check(col.TryUnparse(sb));
+                ctx.SaveString(sb.ToString());
+            }
+
+            ctx.SaveString(_subLoader.ToString());
```

Comment:
This usage of a subcomponent not going to work reliably. The reason why we have data models for loaders is so we can re-open them on new files, while being absolutely assured that the schema is the same as when it was first created. You're assuming, I think, that the schema for every instantiation of a subcomponent loader will be the same, but this is assuredly not guaranteed... under default settings of the `TextLoader`, for example, the `Features` column produced by default has length that depends on what was observed on the file... which of course can change from individual file to file.

For this reason, you will need to change this code to not instantiate from a subcomponent (which has no guarantees), but to instead instantiate from a model (which does have guarantees). You will save the loader's data model to the repository. I see you're already familiar with the general practice of saving models from your treatment of `_pathParser`, so I will not explain it. Generally the saving *might* resemble the code we see [here](https://github.com/dotnet/machinelearning/blob/c023727b76970ab913ec1ce38276508835c17bcf/src/Microsoft.ML.Data/Commands/DataCommand.cs#L399) except instead you'll probably be saving to some binary stream you open in the repository.

The loading back is trickier, since you're not just loading when the repository is open, but many times during the lifetime of that object; so you will have to preserve it somehow. The easiest way I can think of to do that is to just, during loading, copy that stream you wrote over to a `MemoryStream`, then whenever you need to instantiate a new loader (which you currently do using `_subLoader.CreateInstance`), pass the stream and the `IMultiFileSource` over to `ModelFileUtils.LoadPipeline`, over this stream.

Reply:
I understand your concern. Let me look into this further. At the moment, the loading differences were hopefully mitigated by checking that incremental file loads match the schema of the first file loaded. I will look into your suggest to see if there is an alternative.

Reply:
OK great. Just to reinforce the point a bit more strongly, you **cannot** do what is currently being done, for several reasons I'll enumerate here:

1. Our models are intended to be backwards compatible. We make absolutely no assumptions about the same for our command line interface. Any serialization scheme that depends on the command line remaining the same.

2. There is no guarantee that the user will even *use* the command line. The `SubComponent` is specific to command line parsing *only*. Indeed, in this github repository, the command line `maml.exe` simply has not been migrated yet. (In particular, the `Microsoft.ML` API that this is this repos programmatic user interface has no concept of our command line parser, nor `SubComponent`s.)

Let me know if you need advice, I'm happy to help!

---
In reply to: [187676508](https://github.com/dotnet/machinelearning/pull/61#discussion_r187676508) [](ancestors = 187676508)

Reply:
Moved the subLoader to byte array as you suggested. Nice call!

Reply:
Yup... annoying, but necessary. Thanks Tyler!

---
In reply to: [188433067](https://github.com/dotnet/machinelearning/pull/61#discussion_r188433067) [](ancestors = 188433067)

Reply:
Resolved.